### PR TITLE
Codechange: Add unit tests for tile creation

### PR DESCRIPTION
**Pull request by diekmann**
Create unit tests for the `MakeXXX(Tile t, ...` functions.

Also test that the bits which are free according to the documentation are initialized to zero.

## Motivation / Problem

I was trying some vibe coding with GitHub Copilot. Unfortunately, the LLMs I used (GPT 4.1 and Sonnet 3.5) did not get very far. In fact, they got really easily confused about the Tile layout, trying to change the bit meanings (which would break old savegames) or are hallucinating that a tile can be all types at once, ...
This forced me to think myself and I thought I'm putting this thinking into a nice set of unit tests.

The code here was mostly written by hand. LLMs helped with test stubs, but I figured I need to write the bit tests by hand, since I will have to validate them anyway manually.

## Description

Just some unit tests.

### Copilot AI Pre-Review

I tried a few more Copilot LLMs and Gemini 2.5 Pro looked like it can parse the table in docs/landscape_grid.html and extract the free bits. It created a plausible-looking summary for me and I asked it to cross-check this summary against my unit tests. It concluded that they match.
While this commit is NOT vibe coded and NOT written by AI, I'm very happy that an LLM could check that I did not mis-type some of the bits.

### Tested

```sh
$ /usr/bin/ctest -j4 -C Debug -T test --output-on-failure -R "^Make"
Test project /home/vscode/git/OpenTTD/build
      Start  88: MakeTile - MakeClear
      Start  89: MakeTile - MakeField
      Start  90: MakeTile - MakeRailNormal
      Start  91: MakeTile - MakeRailDepot
 1/21 Test  #88: MakeTile - MakeClear .............   Passed    0.13 sec
      Start  92: MakeTile - MakeRoadNormal
 2/21 Test  #89: MakeTile - MakeField .............   Passed    0.13 sec
      Start  93: MakeTile - MakeRoadDepot
 3/21 Test  #90: MakeTile - MakeRailNormal ........   Passed    0.13 sec
      Start  94: MakeTile - MakeHouseTile
 4/21 Test  #91: MakeTile - MakeRailDepot .........   Passed    0.12 sec
      Start  95: MakeTile - MakeTree
 5/21 Test  #93: MakeTile - MakeRoadDepot .........   Passed    0.08 sec
      Start  96: MakeTile - MakeStation
 6/21 Test  #92: MakeTile - MakeRoadNormal ........   Passed    0.10 sec
      Start  97: MakeTile - MakeAirport
 7/21 Test  #95: MakeTile - MakeTree ..............   Passed    0.06 sec
      Start  98: MakeTile - MakeOilrig
 8/21 Test  #94: MakeTile - MakeHouseTile .........   Passed    0.13 sec
      Start  99: MakeTile - MakeWater
 9/21 Test  #96: MakeTile - MakeStation ...........   Passed    0.09 sec
      Start 100: MakeTile - MakeCanal
10/21 Test  #98: MakeTile - MakeOilrig ............   Passed    0.07 sec
      Start 101: MakeTile - MakeShore
11/21 Test  #97: MakeTile - MakeAirport ...........   Passed    0.10 sec
      Start 102: MakeTile - MakeLock
12/21 Test  #99: MakeTile - MakeWater .............   Passed    0.08 sec
      Start 103: MakeTile - MakeShipDepot
13/21 Test #103: MakeTile - MakeShipDepot .........   Passed    0.05 sec
      Start 104: MakeTile - MakeIndustry
14/21 Test #100: MakeTile - MakeCanal .............   Passed    0.10 sec
      Start 105: MakeTile - MakeRailTunnel
15/21 Test #101: MakeTile - MakeShore .............   Passed    0.10 sec
      Start 106: MakeTile - MakeRoadTunnel
16/21 Test #102: MakeTile - MakeLock ..............   Passed    0.14 sec
      Start 107: MakeTile - MakeBridgeRamp
17/21 Test #105: MakeTile - MakeRailTunnel ........   Passed    0.06 sec
      Start 108: MakeTile - MakeObject
18/21 Test #104: MakeTile - MakeIndustry ..........   Passed    0.08 sec
19/21 Test #106: MakeTile - MakeRoadTunnel ........   Passed    0.10 sec
20/21 Test #107: MakeTile - MakeBridgeRamp ........   Passed    0.08 sec
21/21 Test #108: MakeTile - MakeObject ............   Passed    0.08 sec

100% tests passed, 0 tests failed out of 21

Total Test time (real) =   0.59 sec
```

(timings on an old i5-6300U with disabled turbo boost)

## Limitations

There is no formal process in place to make sure docs/landscape_grid.html and src/tests/tile_creation.cpp remain in sync.

Also, the tests reveal that several of the free bits are not actually set to 0. I marked them with `FIXME`.

I think all the bits which are documented to be free should be explicitly set to 0 on tile creation. Otherwise, an author who wants to add a new feature and needs a free bit may assume that the bits are not 0 for legacy savegames, which could cause issues. But this is probably a future PR.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/tests/tile_creation.cpp [hunk 1]
```diff
@@ -0,0 +1,437 @@
+/**
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:1]: Missing the OpenTTD project header:

```
/*
 * This file is part of OpenTTD.
 * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
 * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
 */
```

File: src/tests/tile_creation.cpp [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+#include "../depot_map.h"
+
+#include "../safeguards.h"
+
+
+/**
+ * Helper function to create a tile in a map for testing.
+ * All fields are initialized to ones, so the tests can verify that the Make* functions cleanly initialize all fields.
+ */
+static Tile taintedMockTile()
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:30]: We use Pascal case, never Dromedary case. 🙂 

```suggestion
static Tile TaintedMockTile()
```

**Comment by PeterN** [comment]
Tainted by any use of AI/LLM. Don't.

**Comment by diekmann** [comment]
> Tainted by any use of AI/LLM. Don't.

Wait! I clearly disclosed any use of AI here. And the only code _written_ by an LLM was the test _stubs_. Like

```
TEST_CASE("MakeTile - MakeClear")
{
	XXX (I wrote the initialization by hand, the AI did not get it)

	MakeClear(t, CLEAR_GRASS, 2);

	CHECK(something (the AI mostly left this blank));
	
	REQUIRE(something. I wrote all the bit tests by hand);
}
```

I did spend a fair amount of time to write all those tests by hand :smile: Once written, I used the AI/LLM the *review* the change, which I think is a quite fair use to improve the quality and double check before sending out a PR.

---

I did disclose that I was AI and vibe coding on the OpenTTD codebase. I did this for fun and experimentation. I did not submit a vibe coded PR. And TBH, all the vibe coding I did (with the GitHub copilot free models), did not nearly create code of the quality which is worth a PR. And I will not AI/LLM spam you with low quality PRs :smile:. I promise! As disclosed, this vibe coding motivated me to write the tests in this PR, because the AI/LLM was really bad and stumbled over those things. And seeing and debugging the misconceptions the AI had, motivated me to write those tests (by hand). I actually do think that this was a good use of AI.

Fun fact, in my first vibe coding attempt, with Sonnet 3.5, I asked it to implement diagonal tunnels. It failed at so many things and then the AI blamed crashes on unrelated parts of the codebases and made even worse fixes, so I helped the AI by telling that the game ran fine and all unit tests passed _before_ the AI touched it and asked the AI to correlate the game crashes to its changes and after several hours of letting the AI iterate on that, the AI had reverted all of its changes, except for whitespace mistakes. :rofl:

But seriously, taking one step back, my takeaway was that improving test coverage could really be something good for OpenTTD. So I did spend time on doing this seriously.

**Comment by LordAro** [comment]
There are definitely issues with this PR that will need thorough review, but I don't see it as irredeemable

File: src/tests/tile_creation.cpp [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+	CHECK(IsRoadDepot(t));
+	CHECK(GetDepotIndex(t) == DepotID(1));
+
+	// Check that unused bits are cleared according to docs/landscape_grid.html
+	//REQUIRE(GB(t.m1(), 5, 3) == 0); FIXME: not cleared.
+	REQUIRE(GB(t.m3(), 0, 4) == 0);
+	REQUIRE(GB(t.m4(), 6, 2) == 0);
+	REQUIRE(GB(t.m5(), 2, 4) == 0);
+	REQUIRE(GB(t.m5(), 6, 2) == 2); // not free, but hard-coded to 10b
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:159]: Value and comment should match.

File: src/tests/tile_creation.cpp [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+	REQUIRE(GB(t.m7(), 6, 2) == 0);
+	REQUIRE(GB(t.m8(), 0, 6) == 0);
+	REQUIRE(GB(t.m8(), 12, 4) == 0);
+}
+
+TEST_CASE("MakeTile - MakeHouseTile")
+{
+	Tile t = taintedMockTile();
+	MakeClear(t, CLEAR_GRASS, 2); // A house needs to be placed on a clear tile.
+	MakeHouseTile(t, TownID(8), 1, 1, HouseID(42), 0, false);
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:171]: Out of curiosity, where do these houses and town IDs come from? Randomly chosen?

There are a lot of magic numbers in this PR by definition (because it's low-level looking at map array storage) so any opportunity to use existing constants would be helpful.

File: src/tests/tile_creation.cpp [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+	MakeShipDepot(t, OWNER_NONE, DepotID(1), DEPOT_PART_SOUTH, AXIS_X, WATER_CLASS_SEA);
+
+	CHECK(IsShipDepot(t));
+	CHECK(GetDepotIndex(t) == DepotID(1));
+
+	// Check that unused bits are cleared according to docs/landscape_grid.html
+	REQUIRE(GB(t.m3(), 1, 7) == 0);
+	REQUIRE(t.m4() == 0);
+	REQUIRE(GB(t.m5(), 2, 2) == 0);
+	REQUIRE(GB(t.m5(), 4, 2) == 3); // hard-coded to 1
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:339]: Value and comment should match.

File: src/tests/tile_creation.cpp [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+	CHECK(IsRailDepot(t));
+	CHECK(GetDepotIndex(t) == DepotID(1));
+
+	// Check that unused bits are cleared according to docs/landscape_grid.html
+	//REQUIRE(GB(t.m1(), 5, 2) == 0); FIXME: not cleared.
+	REQUIRE(t.m3() == 0);
+	REQUIRE(GB(t.m4(), 4, 4) == 0);
+	REQUIRE(GB(t.m5(), 2, 2) == 0);
+	REQUIRE(GB(t.m5(), 5, 1) == 0);
+	REQUIRE(GB(t.m5(), 6, 2) == 3); // not free, but hard-coded to 1
```
Comment (2TallTyler) [inline src/tests/tile_creation.cpp:119]: Value and comment should match.

**Comment by 2TallTyler** [comment]
I have not sat down with the tile docs to actually make sure the values in these tests are correct,  but from a read-through of the code here are some things to fix. 🙂

**Comment by PeterN** [comment]
> > Tainted by any use of AI/LLM. Don't.
> 
> Wait! I clearly disclosed any use of AI here. And the only code _written_ by an LLM was the test _stubs_. Like

So what? You still used it.

Good bye.

File: src/tests/tile_creation.cpp (hunk 3 above)
Comment (glx22) [reply src/tests/tile_creation.cpp:159]: Well decimal matches binary.

Comment (2TallTyler) [reply src/tests/tile_creation.cpp:159]: Right, but they should be written in the same format for readability, IMO. 🙂

**Comment by diekmann** [comment]
I assume this PR is now **blocked** on the upcoming AI policy.

While it's still fresh in my head, let me disclose any AI/LLM usage in this commit.

Code written by AI

* Test stubs
* Auto-completion of the `_map.h` `include`s

Further use of AI

* Review of the changes and double-checking for typos in the bit `REQUIRE` tests
* Generic vibe coding against the OpenTTD codebase. This happened in a separate branch and none of that vibe coding leaked into this CL. Because of the quality issues of the resulting vibe code, there obviously is no PR for that vibe code. (In case someone cares, 4ea8a50314336bf9b9889d4f953d5c507a6ead6c was the best result I got. It's not good)

Code not written by AI, but by me

* Both the idea and the implementation of `taintedMockTile`.
* The `REQUIRE` bit tests, which took most of the work.
* The overall research which functions to test.
  * Checking my shell history, I see `find ./src/ -regex '.*\.\(cpp\|hpp\|h\)' | xargs grep -E 'void Make[a-zA-Z0-9]*\(Tile t,'`
* The CMake test inclusion.
* The `safeguards.h` include and how to structure the `include`s, borrowed from `tilearea.cpp`.


So all of the important parts are not "tainted" by any AI/LLM and could be reused.

**Comment by ahyangyi** [comment]
I think it would help to create a clean-house version of this PR.

Not only clean-house is the only way to stay in conform to the AI policy, but also I think that test cases that nobody understands are going to be not useful -- what can one do when a change "breaks" a test case?

**Comment by LordAro** [comment]
If nothing else, it might be worth actually fixing the bugs that this test has found (one tile type at a time)

Assuming that it's something that should be fixed, I'm not actually sure...

**Comment by ahyangyi** [comment]
I would just, by default, assume that's AI hallucination rather than real bugs.

It is a bit hard to believe that the AI, which is unable to play the game at all, has any understanding about what is a bug and what is not.

Of course, if a clean room version that's actually understandable also exposes an existing bug, it's a plus.

**Comment by diekmann** [comment]
Non-AI PR: https://github.com/OpenTTD/OpenTTD/pull/14548
